### DISCUSSION TITLE
Rewrite as a MSBuild task

### DISCUSSION THREAD

**Discussion by bruno-garcia** [microsoft/coyote#403] [Ideas]
I wonder if it was considered to add a rewrite step as a MSBuild target in the NuGet package. This way calling the tool `coyote rewrite` wouldn't be needed as that would be done automatically once the assembly is generated.

Could possibly be done if a specific property is provided like `/p:CoyoteRewrite=true` or be an opt-out when `Configuration=Debug`.

**Comment by pdeligia** [comment]
Thanks a lot for the great suggestion @bruno-garcia!

Internally, quite a few teams are using `coyote rewrite` through a post build task in their `csproj` files, as you are also suggesting, but through their own scripts.

I like the idea to make something generic and provide through the NuGet package, and be optional, as you say. Btw, one common use case that I have seen is that only building a test project triggers rewriting (including rewriting "production" assemblies that the test project depends on), as you only want to rewrite binaries for testing with Coyote and not for deployment. Some internal teams have written scripts that automate this, and these scripts are invoked via a post build task.

Currently, the `coyote rewrite` command line tool accepts both a binary and a [JSON file](https://microsoft.github.io/coyote/#get-started/using-coyote/#configuration) containing a set of binaries and configurations, and it would be great if this task supported both, as the later is especially useful for the scenario that I describe above.

We have been thinking to make such a generic task and include it in the NuGet package, but have not got into this yet. I am adding the `good first issue` tag, as we would love to receive contributions on this! (It also does not require much knowledge of the internals so it should be an approachable issue to tackle.)

**Reply by pdeligia** [reply]
Just to add (to this old topic) that our current recommendation is that someone simply creates their own post-build task (which is typically straightforward to do), instead of us trying to bundle such a task in our NuGet packages. An example how to do so is given in the documentation here: https://microsoft.github.io/coyote/#concepts/binary-rewriting/
